design/aws-ecs-service-discovery-integration.md

```diff
@@ -0,0 +1,235 @@
+# AWS ECS - Support for AWS Cloud Map (Service Discovery) Integration
+
+To address issue [#1554](https://github.com/awslabs/aws-cdk/issues/1554), the
+ECS construct library should provide a way to set
+[`serviceRegistries`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_CreateService.html#ECS-CreateService-request-serviceRegistries)
+on their ECS service at the L1 level.
+
+
+## General approach
+Rather than having the customer instantiate Cloud Map constructs directly and
+thus be forced to learn the API of a separate construct library, we should
+allow the customer to pass in the domain name and other configuration minimally
+needed to instantiate a Cloud Map namespace within an ECS cluster and create a
+Cloud Map service for each ECS service.
+
+The current proposal is to add a method on the ECS Cluster construct,
+`addNamespace` that would take a set of properties that include the namespace
```

Comment:
Why `addNamespace` and not constructor parameters?

Reply:
I suppose that is possible. I thought it might be worth making it explicit, since not everyone will want service discovery in their cluster but having an optional constructor parameter would achieve the same thing.

Reply:
Referring to a [previous comment](https://github.com/SoManyHs/aws-cdk/pull/4#discussion_r255429363), there is also still the possibility of adding more than one namespace to a cluster, although the assumption is that that would be unlikely. Not sure then if it would be better to keep a public `addNamespace()` method or move this to the constructor.

Reply:
Maybe, but this declaration:

```
   private _serviceDiscoveryNamespace?: cloudmap.INamespace;
```

Makes it impossible to have multiple namespaces anyway.

Which right now requires a check in `addNamespace()` by the way to ensure it's only set once.

Again, we avoid all this design-gnashing we don't know the right answers to by simplifying the API down to the bare minimum.

Reply:
Right; I mention the more than one namespace issue here: https://github.com/awslabs/aws-cdk/pull/2065/files#diff-3d958921780fc2ee31330432c62f135eR20

You're right though that if we're not clear about the right answers we could avoid persisting the namespace at the cluster level (i.e. get rid off `addNamespace`) and pass INamespace to `enableServiceDiscovery()`; I do feel that the latter method does add more value however.